### PR TITLE
SDI-139 Create dry-run polling batch job in cmd-api

### PR DESCRIPTION
**Pull request by steverendell**

### CODE REVIEW

File: helm_deploy/values-preprod.yaml [hunk 1]
```diff
@@ -6,11 +6,11 @@ generic-service:
 
   resources:
     requests:
-      cpu: 80m
+      cpu: 150m
       memory: 1000Mi
     limits:
       cpu: 2000m
-      memory: 2000Mi
+      memory: 2500Mi
```
Comment (steverendell) [inline helm_deploy/values-preprod.yaml:13]: I noticed a couple of out of memory crashes recently so upped this a little. Also cpu requests level dictates when the auto-scalar fires up (which we probably want to drop entirely, but for now, just discouraging it  from intervening)

File: src/main/java/uk/gov/justice/digital/hmpps/cmd/api/client/CsrClient.kt [hunk 2]
```diff
@@ -95,39 +122,29 @@ class CsrClient(
...
 
-  @JsonProperty("detailEnd")
   var detailEnd: LocalDateTime,
 
-  @JsonProperty("activity")
   var activity: String
 )
 
 data class CsrModifiedDetailDto @JsonCreator constructor(
+  val id: Long? = null,
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/digital/hmpps/cmd/api/client/CsrClient.kt:135]: the id in the new CMD_NOTIFICATION polling table

File: src/main/java/uk/gov/justice/digital/hmpps/cmd/api/config/ResourceServerConfiguration.kt [hunk 3]
```diff
@@ -68,3 +84,9 @@ class ResourceServerConfiguration : WebSecurityConfigurerAdapter() {
     override fun getPrincipal(): Any = name
   }
 }
+
+@Configuration
+@Profile("!test") // prevent scheduler running during integration tests
+@EnableScheduling
+@EnableSchedulerLock(defaultLockAtLeastFor = "PT1M", defaultLockAtMostFor = "PT1H")
+class SpringSchedulingConfig
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/digital/hmpps/cmd/api/config/ResourceServerConfiguration.kt:92]: moved the EnableScheduling annotation so we can turn it off

File: src/main/java/uk/gov/justice/digital/hmpps/cmd/api/repository/DryRunNotificationRepository.kt [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+  fun countAllByQuantumIdIgnoreCaseAndDetailStartAndParentTypeAndActionType(
+    quantumId: String,
+    detailStart: LocalDateTime,
+    shiftType: ShiftType,
+    actionType: DetailModificationType
+  ): Int
+
+  fun deleteAllByShiftModifiedBefore(shiftModified: LocalDateTime): Int
+}
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/digital/hmpps/cmd/api/repository/DryRunNotificationRepository.kt:30]: this repository and JPA entity class are just copies of the originals

File: src/main/java/uk/gov/justice/digital/hmpps/cmd/api/service/DryRunNotificationService.kt [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+
+  @Transactional(propagation = Propagation.NEVER)
+  fun dryRunSendNotifications() {
+    val start = System.currentTimeMillis()
+    val unprocessedNotifications = dryRunNotificationRepository.findAllByProcessedIsFalse()
+    log.info("dryRunSendNotifications: Sending notifications, found: ${unprocessedNotifications.size}")
+    unprocessedNotifications.groupBy { it.quantumId }
+      .forEach { group ->
+        try {
+          // NO! sendNotification(group.key, group.value)
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/digital/hmpps/cmd/api/service/DryRunNotificationService.kt:47]: actual sending via gov-notify disabled

File: src/main/java/uk/gov/justice/digital/hmpps/cmd/api/service/DryRunNotificationService.kt [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+    val usersWithoutRecentActivity = details
+      .groupingBy { it.quantumId }
+      .aggregate(::latestShiftModified)
+      .filter { it.value.shiftModified.isBefore(cutoffTime) }
+
+    // Now omit details for users with recent changes
+    val detailsToProcess = details.filter { it.quantumId in usersWithoutRecentActivity.keys }
+
+    val allNotifications = detailsToProcess
+      .distinctBy { it.copy(id = null, shiftModified = LocalDateTime.MIN) } // omit id in comparison
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/digital/hmpps/cmd/api/service/DryRunNotificationService.kt:88]: I'll update comment as the modified timestamp also omitted

File: src/main/java/uk/gov/justice/digital/hmpps/cmd/api/service/ShiftService.kt [hunk 7]
```diff
@@ -28,6 +29,8 @@ class ShiftService(
     val start = fromParam.orElse(LocalDate.now(clock))
     val end = toParam.orElse(LocalDate.now(clock))
 
+    log.info("getDetailsForUser: getting for User ${authenticationFacade.currentUsername}, $start - $end")
```
Comment (steverendell) [inline src/main/java/uk/gov/justice/digital/hmpps/cmd/api/service/ShiftService.kt:32]: added to help analyse cache hits / misses

File: src/test/java/uk/gov/justice/digital/hmpps/cmd/api/service/NotificationServiceTest_Generate_Shift.kt [hunk 8]
```diff
@@ -77,13 +77,13 @@ internal class NotificationServiceTest_Generate_Shift {
...
-        end,
-        task,
-        DetailModificationType.ADD
+        quantumId = quantumId,
+        shiftModified = today,
+        shiftType = shiftType,
+        detailStart = start,
+        detailEnd = end,
+        activity = task,
+        actionType = DetailModificationType.ADD
```
Comment (steverendell) [inline src/test/java/uk/gov/justice/digital/hmpps/cmd/api/service/NotificationServiceTest_Generate_Shift.kt:86]: I was going to edit subsequent occurrences like this in a similar way, but realised this test class will eventually get thrown away in favour of the dryrun version

**Review by andymarke — APPROVED** [review]
LGTM

**Pull request merged**